sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/Core/EventHubsExceptionExtensions.cs

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Azure.Messaging.EventHubs.Core
+{
+    /// <summary>
+    ///   The set of extension methods for the <see cref="EventHubsException" /> class.
+    /// </summary>
+    ///
+    internal static class EventHubsExceptionExtensions
+    {
+        /// <summary>
+        ///   Gets the data value related to the exception <see cref="EventHubsException.FailureReason" />.
+        /// </summary>
+        ///
+        /// <param name="instance">The instance that this method was invoked on.</param>
+        ///
+        /// <returns>The Data value or null.</returns>
+        ///
+        public static T GetFailureReasonData<T>(this EventHubsException instance) where T: class =>
```

Comment:
Clever.   I think this is great, but it does require that we tie things to a concrete failure reason.   To avoid adding members to that public enum, it may be worth an overload for string-based keys as well.

Reply:
Thanks for the feedback. The main goal behind associating it with a failure reason was to establish some notion of conflict avoidance and determinism for key values. I considered const string key values as well but I didn't have any great ideas for where those strings would be defined or how they'd be deterministically associated with a given scenario. 

Thoughts?

Reply:
I think that the concept is perfect.... this is the data associated with the failure.   Having a set of consts for the key is probably the way that I'd go, especially if we're considering adding additional metadata about the operation and such.    

That said, I think you're right in that we don't need these methods for any freeform things;  the raw member on `Exception` is fine.   Having additional members here for well-known items is likely a better path than my original thought.